File: backend/src/routes/gridoperator.ts
import { Router, Request, Response } from 'express';
import { lookupGridOperator } from '../services/geocoding';

const router = Router();

router.get('/', async (req: Request, res: Response) => {
  const lat = parseFloat(req.query.lat as string);
  const lng = parseFloat(req.query.lng as string);

  if (isNaN(lat) || isNaN(lng)) {
    return res.status(400).json({ error: 'lat and lng query params are required' });
  }

  try {
    const operator = await lookupGridOperator(lat, lng);
    if (!operator) return res.status(404).json({ error: 'Grid operator not found for this location' });
    return res.json(operator);
  } catch (err) {
    console.error('Grid operator lookup error:', err);
    return res.status(500).json({ error: 'Lookup failed' });
  }
});

export default router;

File: backend/app/routers/gridoperator.py
"""GET /api/gridoperator — port of routes/gridoperator.ts."""
from __future__ import annotations

from fastapi import APIRouter, HTTPException, Query

from app.models import GridOperator
from app.services.geocoding import lookup_grid_operator

router = APIRouter()


@router.get("", response_model=GridOperator)
async def get_grid_operator(
    lat: float = Query(..., description="Latitude"),
    lng: float = Query(..., description="Longitude"),
) -> GridOperator:
    try:
        operator = await lookup_grid_operator(lat, lng)
    except Exception as exc:
        import logging
        logging.getLogger(__name__).error("Grid operator lookup error: %s", exc)
        raise HTTPException(status_code=500, detail="Lookup failed")

    if not operator:
        raise HTTPException(status_code=404, detail="Grid operator not found for this location")
    return operator
